turtle_tf2_cpp/src/turtle_tf2_message_filter.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <geometry_msgs/msg/point_stamped.hpp>
+#include <message_filters/subscriber.h>
+
+#include <rclcpp/rclcpp.hpp>
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/create_timer_ros.h>
+#include <tf2_ros/message_filter.h>
+#include <tf2_ros/transform_listener.h>
+#ifdef TF2_CPP_HEADERS
+  #include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
+#else
+  #include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#endif
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+class PoseDrawer : public rclcpp::Node
+{
+public:
+  PoseDrawer()
+  : Node("turtle_tf2_pose_drawer")
+  {
+    // Declare and acquire `target_frame` parameter
+    this->declare_parameter<std::string>("target_frame", "turtle1");
+    this->get_parameter("target_frame", target_frame_);
+
+    typedef std::chrono::duration<int> seconds_type;
+    seconds_type buffer_timeout(10);
+
+    tf2_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
+    // set timer interface before call to waitForTransform,
+    // to avoid a tf2_ros::CreateTimerInterfaceException is thrown.
+    auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+      this->get_node_base_interface(),
+      this->get_node_timers_interface());
+    tf2_buffer_->setCreateTimerInterface(timer_interface);
+    tf2_listener_ =
+      std::make_shared<tf2_ros::TransformListener>(*tf2_buffer_);
+    point_sub_.subscribe(this, "/turtle3/turtle_point_stamped");
+    tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
+      point_sub_, *tf2_buffer_, target_frame_, 100, this->get_node_logging_interface(),
+      this->get_node_clock_interface(), buffer_timeout);
+    tf2_filter_->registerCallback(&PoseDrawer::msgCallback, this);
```

Comment:
I think it would be nice to have some empty lines between separate tools, like:
```suggestion
    tf2_buffer_->setCreateTimerInterface(timer_interface);
    tf2_listener_ =
      std::make_shared<tf2_ros::TransformListener>(*tf2_buffer_);

    point_sub_.subscribe(this, "/turtle3/turtle_point_stamped");
    tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
      point_sub_, *tf2_buffer_, target_frame_, 100, this->get_node_logging_interface(),
      this->get_node_clock_interface(), buffer_timeout);
    tf2_filter_->registerCallback(&PoseDrawer::msgCallback, this);
```

Reply:
@kurshakuz  Yes, you're right. I added an empty line to separate them.

Reply:
@kurshakuz  By the way, really it seems that we didn't use the tf2_geometry_msgs, but I tested if I remove it, it will get wrong when building the cpp package. And in the ROS1 version of this tutorial, there also had this header, so I think we need it.

Reply:
The error looks like this:
try_tutorials/turtle_tf2_cpp$ colcon build --symlink-install
Starting >>> turtle_tf2_cpp
--- stderr: turtle_tf2_cpp                                
/usr/bin/ld: CMakeFiles/turtle_tf2_message_filter.dir/src/turtle_tf2_message_filter.cpp.o: in function `geometry_msgs::msg::PointStamped_<std::allocator<void> >& tf2_ros::BufferInterface::transform<geometry_msgs::msg::PointStamped_<std::allocator<void> > >(geometry_msgs::msg::PointStamped_<std::allocator<void> > const&, geometry_msgs::msg::PointStamped_<std::allocator<void> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::chrono::duration<long, std::ratio<1l, 1000000000l> >) const':
turtle_tf2_message_filter.cpp:(.text._ZNK7tf2_ros15BufferInterface9transformIN13geometry_msgs3msg13PointStamped_ISaIvEEEEERT_RKS7_S8_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENSt6chrono8durationIlSt5ratioILl1ELl1000000000EEEE[_ZNK7tf2_ros15BufferInterface9transformIN13geometry_msgs3msg13PointStamped_ISaIvEEEEERT_RKS7_S8_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENSt6chrono8durationIlSt5ratioILl1ELl1000000000EEEE]+0x5a): undefined reference to `std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > > tf2::getTimestamp<geometry_msgs::msg::PointStamped_<std::allocator<void> > >(geometry_msgs::msg::PointStamped_<std::allocator<void> > const&)'
/usr/bin/ld: turtle_tf2_message_filter.cpp:(.text._ZNK7tf2_ros15BufferInterface9transformIN13geometry_msgs3msg13PointStamped_ISaIvEEEEERT_RKS7_S8_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENSt6chrono8durationIlSt5ratioILl1ELl1000000000EEEE[_ZNK7tf2_ros15BufferInterface9transformIN13geometry_msgs3msg13PointStamped_ISaIvEEEEERT_RKS7_S8_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENSt6chrono8durationIlSt5ratioILl1ELl1000000000EEEE]+0x7a): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > tf2::getFrameId<geometry_msgs::msg::PointStamped_<std::allocator<void> > >(geometry_msgs::msg::PointStamped_<std::allocator<void> > const&)'
/usr/bin/ld: turtle_tf2_message_filter.cpp:(.text._ZNK7tf2_ros15BufferInterface9transformIN13geometry_msgs3msg13PointStamped_ISaIvEEEEERT_RKS7_S8_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENSt6chrono8durationIlSt5ratioILl1ELl1000000000EEEE[_ZNK7tf2_ros15BufferInterface9transformIN13geometry_msgs3msg13PointStamped_ISaIvEEEEERT_RKS7_S8_RKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENSt6chrono8durationIlSt5ratioILl1ELl1000000000EEEE]+0xcf): undefined reference to `void tf2::doTransform<geometry_msgs::msg::PointStamped_<std::allocator<void> > >(geometry_msgs::msg::PointStamped_<std::allocator<void> > const&, geometry_msgs::msg::PointStamped_<std::allocator<void> >&, geometry_msgs::msg::TransformStamped_<std::allocator<void> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/turtle_tf2_message_filter.dir/build.make:192：turtle_tf2_message_filter] 错误 1
make[1]: *** [CMakeFiles/Makefile2:115：CMakeFiles/turtle_tf2_message_filter.dir/all] 错误 2
make: *** [Makefile:141：all] 错误 2
---
Failed   <<< turtle_tf2_cpp [17.0s, exited with code 2]

Summary: 0 packages finished [17.2s]
  1 package failed: turtle_tf2_cpp
  1 package had stderr output: turtle_tf2_cpp

Reply:
And I increased the queue size from 10 to 100:

tf2_filter_ = std::make_shared<tf2_ros::MessageFilter<geometry_msgs::msg::PointStamped>>(
      point_sub_, *tf2_buffer_, target_frame_, 100, this->get_node_logging_interface(),
      this->get_node_clock_interface(), buffer_timeout);

because if the queue size is small, there will have some warning outputs:

message filter dropping message: frame 'world' at time 1632119254.938 for reason 'discarding message because the queue is full'

Reply:
@kurshakuz And I also remove this line under the includings:
using std::placeholders::_1;
I test we don't need that line, and it can still work as expected.

Reply:
@kurshakuz I modified the cpp file according to your last nitpicks suggestion.

Reply:
Thank you! Looks pretty good to me

Reply:
> @kurshakuz By the way, really it seems that we didn't use the tf2_geometry_msgs, but I tested if I remove it, it will get wrong when building the cpp package. And in the ROS1 version of this tutorial, there also had this header, so I think we need it.

Yes, indeed we need a `tf2_geometry_msgs.hpp` because it contains required conversion methods.